feat(forms): allow min/max validator to be bound to null

iRealNirmal opened the pull request:
If the validator is bound o be
ull then no validation occurs and attribute is not added to DOM.

PR already raised for minLength and maxLength valdiator (https://github.com/angular/angular/pull/42565).
Changes were discussed in #42378

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Null data type is not supported

Issue Number: #42267.


## What is the new behavior?
Added support for null data type

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

This is one of many PR which is going to be raised for adding support of null validation. One can go through conversation here
#42378 (comment)
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

iRealNirmal committed:
feat(forms): allow disabling min/max validators dynamically (by setting the value to `null`)

This commit updates the logic of the `min` and `max` validators to allow
disabling them dynamically in case `null` is provided as a value. For example: `<input
type="number" [min]="minValue">`, when `minValue` might be set to `null` in a
component class. This should allow `min` and `max` validators to be used for dynamic forms.

Note: similar support was added to the `minLength` and `maxLength`
validators earlier (see #42565).

iRealNirmal commented:
@AndrewKushnir I will be adding test cases as well once I get go ahead with current implementation.

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -200,6 +203,11 @@ export class MaxValidator extends AbstractValidatorDirective implements OnChange
     this.handleChanges(changes);
   }
+
+  /** @nodoc */
+  enabled(): boolean {
+    return this.max != null /* both `null` and `undefined` */;
+  }
```
I think we should try moving the `enabled` function to the `AbstractValidatorDirective` class, so it may look like this:
```
@Directive()
abstract class AbstractValidatorDirective implements Validator {
  ...
  handleChanges(changes: SimpleChanges): void {
      ...
      this._validator = this.enabled() ? this.createValidator(input) : nullValidator;
      ...
  }
  ...
  // TODO: add docs as needed.
  enabled(): boolean {
    return this[this.inputName] != null /* both `null` and `undefined` */;
  }
}
```
The default implementation of the `enabled` function can just check whether the input is not `null` and `undefined` (that'd work for most of the validators) and for validators that have different logic (like `required`) - the `enabled` function can be extended in a subclass.

@iRealNirmal what do you think?

iRealNirmal replied:
yes that is eventually we have to do, it will be better to do that right now. I will move it to that

iRealNirmal replied:
I have done this changes just  one dilemma here  in  // TODO: add docs as needed. you want me to add comment right?(You don't need to provide comment I just  want to confirm) or you meant it to be added some where else?

AndrewKushnir reviewed (commented):
Thanks for creating this PR @iRealNirmal 👍 

I've left a comment with a proposal to move the `enabled` logic into the base class, please let me know what you think. Thank you.

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -140,6 +140,11 @@ abstract class AbstractValidatorDirective implements Validator {
+
+  // TODO: add docs as needed
+  enabled(): boolean {
+    return this.inputName != null /* both `null` and `undefined` */;
```
I think this change is incorrect since `this.inputName` would contain the name of the input, not its value. The change should look like this:
```suggestion
    return this[this.inputName] != null /* both `null` and `undefined` */;
```

iRealNirmal replied:
yes you are correct. for now I am getting error
`Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'AbstractValidatorDirective'.
  No index signature with a parameter of type 'string' was found on type 'AbstractValidatorDirective'` will update here once it's resolved

iRealNirmal replied:
I was thinking what if I capture value of element in handleChanges and use of it like below, let me know your thoughts @AndrewKushnir 

```
handleChanges(changes: SimpleChanges): void {
    if (this.inputName in changes) {
      this._inputValue = this.normalizeInput(changes[this.inputName].currentValue);
      this._validator = this.createValidator(this._inputValue);
      if (this._onChange) {
        this._onChange();
      }
    }
  }
  
  
  ...
  // TODO: add docs as needed.
  enabled(): any {
    return this._inputValue != null /* both `null` and `undefined` */;
  }
```

AndrewKushnir replied:
@iRealNirmal I think you can use a type cast there:

```
return (this as unknown as {[key: string]: unknown})[this.inputName] != null /* both `null` and `undefined` */;
```

iRealNirmal replied:
yes @AndrewKushnir it helped, I will do the required change. Thanks for helping in it.

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -177,20 +182,23 @@ export const MAX_VALIDATOR: StaticProvider = {
   override normalizeInput = (input: string): number => parseFloat(input);
   /** @internal */
-  override createValidator = (max: number): ValidatorFn => maxValidator(max);
+  override createValidator = (max: number):
```
The `createValidator` function should remain the same here and instead we should update the `handleChanges` function inside the `AbstractValidatorDirective` directive to:
```
  validate(control: AbstractControl): ValidationErrors|null {
    ...
    if (this.enabled()) {
      const input = this.normalizeInput(changes[this.inputName].currentValue);
      this._validator = this.createValidator(input);
    } else {
      this._validator = nullValidator;
    }
    ...
  }
```

With that change we can also make sure that the `normalizeInput` function is not called with unexpected input value (such as `null` or `undefined`), which is not expected by some of the functions, for example (the input is expected to be `string` only):

```
override normalizeInput = (input: string): number => parseFloat(input);
```

iRealNirmal replied:
I have done change, you can  review once you get a chance

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -177,20 +182,23 @@ export const MAX_VALIDATOR: StaticProvider = {
   /** @internal */
   override inputName = 'max';
   /** @internal */
   override normalizeInput = (input: string): number => parseFloat(input);
```
The `number` type is missing here (it's pre-existing, but we can improve this in this PR as well):
```suggestion
  override normalizeInput = (input: string|number): number => parseFloat(input);
```

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -237,20 +245,23 @@ export const MIN_VALIDATOR: StaticProvider = {
   /** @internal */
   override inputName = 'min';
   /** @internal */
   override normalizeInput = (input: string): number => parseFloat(input);
```
```suggestion
  override normalizeInput = (input: string|number): number => parseFloat(input);
```

AndrewKushnir reviewed (commented):
Thanks for additional changes @iRealNirmal, please see my comments below.

>I have done this changes just one dilemma here in // TODO: add docs as needed. you want me to add comment right?(You don't need to provide comment I just want to confirm) or you meant it to be added some where else?

Yes, we need to add docs for that function (in a similar way other `AbstractValidatorDirective ` class methods are documented).

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -177,18 +187,18 @@ export const MAX_VALIDATOR: StaticProvider = {
   override inputName = 'max';
   /** @internal */
-  override normalizeInput = (input: string): number => parseFloat(input);
+  override normalizeInput = (input: string|number): number => toNumber(input);
```
Since the `toNumber` function uses `parseInt` inside, we can not use it here as is (otherwise floating point numbers would not be supported). We should probably rename `toNumber` -> `toInt` and create a similar function that would use `parseFloat` inside (and call it `toFloat`).

iRealNirmal replied:
yes great catch I somehow missed it, I will rename as well, thanks

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -140,6 +140,16 @@ abstract class AbstractValidatorDirective implements Validator {
+   * When value of input is null or undefined return false
+   * else return true
+   */
+  enabled(): boolean {
+    return (this as unknown as {[key: string]: unknown})[this.inputName] !=
+        null /* both `null` and `undefined` */;
+  }
```
Just for readability:
```suggestion
  enabled(): boolean {
    const inputValue = (this as unknown as {[key: string]: unknown})[this.inputName];
    return inputValue != null /* both `null` and `undefined` */;
  }
```

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -140,6 +140,16 @@ abstract class AbstractValidatorDirective implements Validator {
     this._onChange = fn;
   }
+
+  /**
+   * @description
+   * When value of input is null or undefined return false
+   * else return true
+   */
```
```suggestion
  /**
   * @description
   * Determines whether this validator is active or not. Base class implementation
   * checks whether an input is defined (the value is different from `null` and `undefined`).
   * Validator classes that extend this base class can override this function with the logic
   * specific to a particular validator directive.
   */
```

AndrewKushnir reviewed (commented):
@iRealNirmal thanks for additional changes. I've left a few comments.

The general logic looks good, it'd be great to add tests to verify new behavior (based on the tests that you've added for MinLength and MaxLength validators in the previous PR).

iRealNirmal commented:
> test

Yes I will add test case as well, I was just waiting for our code approach to be finalised before writing test cases to prevent rewriting of it. I will add in next commit with suggested changes.

iRealNirmal commented:
@AndrewKushnir can you please  review it again once ?

AndrewKushnir commented:
@iRealNirmal sorry for the delay! I will have a look at the changes within the next day or so and let you know. Thank you.

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -177,18 +199,18 @@ export const MAX_VALIDATOR: StaticProvider = {
   override inputName = 'max';
   /** @internal */
-  override normalizeInput = (input: string): number => parseFloat(input);
+  override normalizeInput = (input: string|number): number => toInteger(input);
```
```suggestion
  override normalizeInput = (input: string|number): number => toFloat(input);
```

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -237,18 +259,18 @@ export const MIN_VALIDATOR: StaticProvider = {
   override inputName = 'min';
   /** @internal */
-  override normalizeInput = (input: string): number => parseFloat(input);
+  override normalizeInput = (input: string|number): number => toInteger(input);
```
I think it should be `toFloat`, not `toInteger`:
```suggestion
  override normalizeInput = (input: string|number): number => toFloat(input);
```
Btw, it's interesting that we don't have failing tests after this change. @iRealNirmal if you get a chance, could you please add a test that would verify that we can not have a regression here (we can probably have min/max and a value in tests as float numbers and come up with a case when `parseInt` would cause an issue)?

iRealNirmal replied:
hi @AndrewKushnir thanks for catching mistake. Regarding your comment
`we can probably have min/max and a value in tests as float numbers and come up with a case when parseInt would cause an issue`

So I doubt this can be valid scenario because if min is `1.0` and max is `3.0` and user enters just `2` then still it will be correct to user always enter decimal value developers should add pattern or regex.

Please feel free tp correct me If am wrong in above scenario or in case you have other thought in your mind.

AndrewKushnir replied:
I was thinking of a scenario where `min="1.5"` (thus `min` becomes `1` after `parseInt` operation) and you try to input `1.6`. With the `parseFloat` this would work, but with the `parseInt` call (which you had in the PR for some time) - validation would not indicate this as a problem. Please let me know if you need more information.

iRealNirmal replied:
You are correct, It's strange what  I saw that we already have test  cases of float number under
`should validate max for float number` but it didn't failed when I changed it toInteger also  I wrote some test cases but still it worked. I have to spend some  more time to find a root cause and fix for this. Feel Free to let me know in case you have something in mind. In case I found some fix or root cause I will update this  thread.

iRealNirmal replied:
@AndrewKushnir reason of it's not failing is in validator method of
maxValidator we are using parseFloat inside file (form/src/validators.ts) let me know what you expect me  to do about it.

AndrewKushnir commented on packages/forms/test/template_integration_spec.ts:
```diff
@@ -2005,6 +2005,91 @@ import {NgModelCustomComp, NgModelCustomWrapper} from './value_accessor_integrat
              verifyValidatorAttrValues({minlength: null, maxlength: null});
              verifyFormState({isValid: true});
            }));
+        it('should not include the min and max validators for null', fakeAsync(() => {
```
Could you please add a new line before a new test?
```suggestion

        it('should not include the min and max validators for null', fakeAsync(() => {
```

iRealNirmal replied:
added

AndrewKushnir reviewed (commented):
@iRealNirmal I've added some comments, could you please have a look when you get a chance? It also seems like there is a merge conflict with the main branch. It'd be helpful to rebase and resolve conflicts, so that we can run extra tests in Google's codebase.

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -13,16 +13,24 @@ import {AbstractControl} from '../model';
 }
 
+/**
+ * Method that updates string to float if not alread a number
```
```suggestion
 * Method that ensures that provided value is a float (and converts it to float if needed).
```

AndrewKushnir commented:
FYI I've also started [tests in Google's codebase](http://test/OCL:395399002:BASE:395399016:1631077976507:68c46715) (internal-only link) to see if there are some use-cases that we'd need to cover.

iRealNirmal commented:
hi @AndrewKushnir any update on this PR ?

AndrewKushnir commented:
@iRealNirmal initial tests in Google's codebase went well, I'll start a more broad set of tests asap and keep you updated on how it goes. Thank you.

AndrewKushnir commented:
[Global presubmit](https://test.corp.google.com/OCL:395399002:BASE:397895447:1632190321151:5027776e) (internal-only link).

AndrewKushnir reviewed (approved):
@iRealNirmal sorry for the delay!

Initial tests in Google's codebase went well, I'm currently re-running them one last time and will let you know.

Overall I think this change is ready for merge and the final update that I'd like to request is to change commit message (that would also become a part of the release notes). Here is my proposal:

```
feat(forms): allow disabling min/max validators dynamically (by setting the value to `null`) 

This commit updates the logic of the `min` and `max` validators to allow disabling them
dynamically in case `null` is provided as a value. For example: `<input type="number" 
[min]="minValue">`, when `minValue` might be set to `null` in a component class. This 
should allow `min` and `max` validators to be used for dynamic forms.

Note: similar support was added to the `minLength` and `maxLength` validators earlier
(see #42565).
```

Thank you.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

jessicajaniuk reviewed (approved):
LGTM :cookie:
reviewed-for: public-api

IgorMinar reviewed (approved):
LGTM

Reviewed-for: public-api

AndrewKushnir commented:
@iRealNirmal just want to let you know that this PR was merged and this feature will be included into Angular v13 release! 🎉 

I also want to take this opportunity to say Thank You for the work that you did in this and other PRs! We made a few iterations on this change and you've addressed all the feedback and was very responsive. Thanks again for all your help!

iRealNirmal commented:
> @iRealNirmal just want to let you know that this PR was merged and this feature will be included into Angular v13 release! 🎉
> 
> I also want to take this opportunity to say Thank You for the work that you did in this and other PRs! We made a few iterations on this change and you've addressed all the feedback and was very responsive. Thanks again for all your help!

@AndrewKushnir thank you, thank you very much for your kind words and constant guidance, it was pleasure and fun working on it. Also am happy to give back community as am using angular since quite a long time and indirectly it helps to understand work being done by framework under the hood.

As next item I will work on remaining validators, also feel free to let me know if there is anything else you want me to take care.

AndrewKushnir commented:
@iRealNirmal thanks for the reply. If you get a chance to help, it'd be great to transition other validators (starting with minlength and maxlength) to use `AbstractValidatorDirective` as a base class, as you mentioned. Please let us know if you need any additional information, we'll be glad to help.

iRealNirmal commented:
sure I will take care of it.
